verifiers/internal/gha/slsaprovenance/v0.2/generic.go

```diff
@@ -0,0 +1,149 @@
+package v02
+
+import (
+	"fmt"
+	"time"
+
+	intoto "github.com/in-toto/in-toto-golang/in_toto"
+	slsa02 "github.com/in-toto/in-toto-golang/in_toto/slsa_provenance/v0.2"
+	serrors "github.com/slsa-framework/slsa-verifier/v2/errors"
+
+	"github.com/slsa-framework/slsa-verifier/v2/verifiers/internal/gha/slsaprovenance/common"
+)
+
+var (
+	goBuilderBuildType = "https://github.com/slsa-framework/slsa-github-generator/go@v1"
+
+	genericGeneratorBuildType   = "https://github.com/slsa-framework/slsa-github-generator/generic@v1"
+	containerGeneratorBuildType = "https://github.com/slsa-framework/slsa-github-generator/container@v1"
+	npmCLIBuildType             = "https://github.com/npm/cli/gha@v1"
+
+	// Legacy build types.
+	legacyGoBuilderBuildType = "https://github.com/slsa-framework/slsa-github-generator-go@v1"
+	legacyBuilderBuildType   = "https://github.com/slsa-framework/slsa-github-generator@v1"
+
+	// genericGHABuildType is used by some tests.
+	// TODO: Update tests to use a real buildType.
+	genericGHABuildType = "https://github.com/Attestations/GitHubActionsWorkflow@v1"
+)
+
+// GenericProvenanceV02 represents SLSA v0.2 provenance for the
+// slsa-github-generator Go builder, generic generator, container generator, and npm CLI.
+type GenericProvenanceV02 struct {
+	prov *intotoAttestation
+}
+
+// Predicate implements ProvenanceV02.Predicate.
+func (p *GenericProvenanceV02) Predicate() slsa02.ProvenancePredicate {
+	return p.prov.Predicate
+}
+
+// BuilderID implements Provenance.BuilderID.
+func (p *GenericProvenanceV02) BuilderID() (string, error) {
+	return p.prov.Predicate.Builder.ID, nil
+}
+
+// SourceURI implements Provenance.SourceURI.
+func (p *GenericProvenanceV02) SourceURI() (string, error) {
+	if len(p.prov.Predicate.Materials) == 0 {
```

Comment:
is there a way we could share all these function with a "base structure", and only re-wrrite those that need changes?